### ISSUE TITLE
VAE not using n_batch?

### ISSUE THREAD

**Issue by thouis** [YosefLab/scVI#268]
I might be missing something, but it looks like the model doesn't use the batch info during encoding:
https://github.com/YosefLab/scVI/blob/bf26369b0a2edd1bb16c3d90524d1e5cf0138bbb/scvi/models/vae.py#L72

Is that intentional?  My understanding of the paper is that it should be used during encoding as well.

**Comment by coh-racng** [comment]
As long as we calculate the local_l_var and local_l_var per batch, that is passed into the forward() function of VAE model during training to calculate kl_divergence. 
 `kl_divergence_l = kl(Normal(ql_m, torch.sqrt(ql_v)), Normal(local_l_mean, torch.sqrt(local_l_var))).sum(dim=1)`

Should the library random variable be a B-dimensional variable? According to Methods, 

> B denotes the number of batches and ℓ𝜇,ℓ𝜎∈ℝ𝐵+ parameterize the prior for the scaling factor (on a log scale). ℓ𝜇,ℓ𝜎  are set to be the empirical mean and variance of the log-library size per batch. We note that the random variable ℓ𝑛 is not the log-library size (scaling the sampled observation) itself but a scaling factor that is expected to correlate strongly with log-library size (hence the choice of the parameters). 

But the l_encoder seems to have a single mean and variance (ql_m and ql_v) when it is initialized in VAE model:
```python 
self.l_encoder = Encoder(n_input, 1, n_layers=1, n_hidden=n_hidden, dropout_rate=dropout_rate)
```
Should we pass  `n_output=n_batch` to make it  B-dimensional? Should l_encoder also take batch_indexes as input during training using `n_cat_list=[n_batch]`? and during sampling in `inference()` with `ql_m, ql_v, library = self.l_encoder(x_, batch_index)`?

**Comment by romain-lopez** [comment]
@thouis Thanks for your message. It is true that the notations in the paper parameterizes the variational posterior for the latent space as q(z | x, z). However, in practice we found it is not useful to the decoder. One explanation is that provided x, it is very easy for the network to tell appart the two batches.

@coh-racng Thanks for your answer. Basically, the prior for the library size has multiple values for batches. However, the variational posterior q(l | x, s) denotes the distribution of a scalar random variable. As for the last remark, it is not necessary in practice to feed the information s in the neural network as it is redundant with x. 

Hope that helps!

**Comment by romain-lopez** [comment]
Closing issue for inactivity

**Comment by coh-racng** [comment]
Why is it redundant with x when x only has raw expression data and no information about batch index? It seems like the batch information fed to the decoder. How does scVI perform when encoder is also fed the batch information?
